testutils/lndir.man
.\" Copyright (c) 1993, 1994  X Consortium
.\" Copyright (C) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2007,
.\" 2008, 2009, 2010, 2001, 2014, 2019, 2024 Free Software Foundation,
.\" Inc.
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a
.\" copy of this software and associated documentation files (the "Software"),
.\" to deal in the Software without restriction, including without limitation
.\" the rights to use, copy, modify, merge, publish, distribute, sublicense,
.\" and/or sell copies of the Software, and to permit persons to whom the
.\" Software furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
.\" THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
.\" WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
.\" OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
.\" SOFTWARE.
.\"
.\" Except as contained in this notice, the name of the X Consortium shall not
.\" be used in advertising or otherwise to promote the sale, use or other
.\" dealing in this Software without prior written authorization from the
.\" X Consortium.
.TH LNDIR 1 "Release 6" "X Version 11"
.SH NAME
lndir \- create a shadow directory of symbolic links to another directory tree
.SH SYNOPSIS
.B lndir
fromdir [todir]
.SH DESCRIPTION
.I Lndir
makes a shadow copy
.I todir
of a directory tree
.I fromdir,
except that the shadow is not
populated with real files but instead with symbolic links pointing at
the real files in the
.I fromdir
directory tree.  This is usually useful for maintaining source code for
different machine architectures.  You create a shadow directory
containing links to the real source which you will have usually NFS
mounted from a machine of a different architecture, and then recompile
it.  The object files will be in the shadow directory, while the
source files in the shadow directory are just symlinks to the real
files.
.PP
This has the advantage that if you update the source, you need not
propagate the change to the other architectures by hand, since all
source in shadow directories are symlinks to the real thing: just cd
to the shadow directory and recompile away.
.PP
The
.I todir
argument is optional and defaults to the current directory.  The
.I fromdir
argument may be relative (e.g., ../src) and is relative to
.I todir
(not the current directory).
.PP
.ft B
.\" CVS.adm is used by the Concurrent Versions System.
Note that RCS, SCCS, and CVS.adm directories are not shadowed.
.ft
.PP
If you add files, simply run
.I lndir
again.  Deleting files is a more painful problem; the symlinks will
just point into never never land.
.SH BUGS
.I Patch
gets upset if it cannot change the files.  You should never run
.I patch
from a shadow directory anyway.
.PP
You need to use something like
.nf
	find todir \-type l \-print | xargs rm
.fi
to clear out all files before you can relink (if fromdir moved, for instance).
Something like
.nf
	 find . \\! \-type d \-print
.fi
will find all files that are not directories.


NEWS
New in CSSC-1.5.0-rc2, 2024-05-13

	 * This release is more careful to detect I/O failures when
           writing to stdout in prs and prt.

         * This release should build on more modern platforms.

	 * The --with-googletest configure option is removed (now we
           always use it).

	 * Updated versions of gnulib and googletest.

New in CSSC-1.4.1, 2019-05-07

         * This release - and future releases - of CSSC must be
           compiled with a C++ compiler that supports the 2011
           C++ standard.

         * When the history file is updated (with admin or delta for
           example) the history file is not made executable simply
           because the 'x' flag is set.  Instead, preserve the
           executable-ness from the history file we are replacing.

         * This release is based on updated versions of gnulib and of
          the googletest unit test framework.

New in CSSC-1.4.0, 2014-07-11

         * New stable release, almost identical to the test release
           1.3.2 - the only changes are version numbering and updates
           to the ChangeLog and NEWS files.  The previous stable
           release was 1.3.0, so the following two sections of this
           file set out what has changed since 1.3.0.

New in CSSC-1.3.2, 2014-06-30

        * This release is based on updated versions of gnulib and of
          the googletest unit test framework (the test framework is of
          course not linked into the installed CSSC binaries, it is
          just used for unit tests).

        * This release uses a new tar format.  Up to now we have used
          the v7 Unix tar format, but this does not cope correctly
          with filenames longer than 99 characters, and we now have
          one of those.  We now use the ustar format defined by
          POSIX 1003.1-1988.  If you need to unpack a CSSC
          distrubition with a version of tar which only deals with the
          v7 format, unpack the distribution somewhere else, rename or
          delete the long files (they are not essential to the build
          process) and re-pack it. This new tar format will likely
          also be used for all future releases.

        * Fix Savannah bug 33633, prs -d":GB:" coredumps when
          expanding %I%.

New in CSSC-1.3.1, 2011-05-10

        * prs now prints ignored deltas when expanding the :DI:
          keyword.  The expansion of :DI: varies between SCCS
          implementations, as described in the manual.  This partly
          fixes Savannah bug #33154: SCCS versions vary wrt ignored
          deltas (that is, this change covers only the prs binary).

        * Fix Savannah bug bug #33153: "prs" includes "AUTO NULL
          DELTA"s.  "prs -r" will no longer print any additional
          deltas that happen to share the same timestamp with the last
          matching delta.

        * If the SID identified by "prs -r" is not found, prs now
          exits with status 1, not 0.  This fixes SourceForge bug
          #1300722.

	* The headers produced by sccsdiff have changed.  A different
          header is used when the -p option is specified and when
          it's not.  The new behaviour matches OpenSolaris 2009.06.

	* The ordering of the ouput of "sccs unedit" is now different;
          the new order matches OpenSolaris 2009.06.

	* "get -L" will now send the ID of the gotten SID to stderr,
          which better matches OpenSolaris 2009.06.  The indentation
          of the delta summary has also been adjusted.

	* To pass an empty list of MRs to admin, you now have to
          provide an argument (as opposed to omitting it).  For
          example, "admin -i/dev/null -fv -m' ' s.foo".  This aligns
          better with OpenSolaris 2009.06 but it is a
          backward-incompatible change.

New in CSSC-1.3.0, 2010-11-07

        * This release is identical to CSSC-1.3.0rc1 apart from the
          version number.

New in CSSC-1.3.0rc1, 2010-05-16

        * The CSSC manual is now published under version 1.3 of the
          GNU Free Documentation License.

 	* The code and build system have been modernised and some
          redundant files have been removed.  This may make it harder
          to build CSSC on very old systems.  However, it's now easier
          to maintain.  If CSSC no longer builds on a system that's
          important to you, please send mail to <bug-cssc@gnu.org>
          explaining your problem.

	* CSSC now uses the GNU portability library, gnulib.

	* CSSC now requires a C++ compiler with working exceptions and
          a working STL implementation.

	* Adoption of some STL data structures has made some parts of
          CSSC faster, notably "val".

	* CSSC now includes unit tests.  We also run the y2k tests for
          "make check".  If valgrind is installed, it will be used to
          for verification during tests.  You can turn this off by
          givig the --without-valgrind option to configure.

	* The configure script now supports a new option
          "--enable-warnings" which enables many GCC warning options.

New in CSSC-1.2.0, 2009-04-11

	* This stable release is almost identical to the 1.1.1 test
          release; there are no functional changes.

New in CSSC-1.1.1, 2008-01-18

	* Updated the NEWS file to give estimated release dates
	  and user-visible changes for ancient releases.

	* The -l and -L options of "get" are now implemented.

	* If you use "admin -dlX" to unlock release X, but all
	  releases are locked, an explanatory error message is issued.
	  Previously, the attempt was simply silently ignored.

	* prs no longer has a -R option.

	* When CSSC tools are passed the name of a directory on the
	  command line, they operate on the SCCS files in that directory.
	  As of this release, subdirectories are ignored (instead of
	  generating an error when CSSC tries to open the directory as
	  a file).

New in CSSC-1.1.0, 2007-12-17


testutils/lndir.c

/* lndir.c; part of GNU CSSC.
 *
 * Copyright (C) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2007,
 * 2008, 2009, 2010, 2011, 2014, 2019, 2024 Free Software Foundation,
 * Inc.
 *
 * This file is derived from xc/config/util/lndir.c in the X11R6 distribution.
 * It's been changed to make use of GNU Autoconf rather than xmkmf (imake)
 * for portability.
 *   -- James Youngman <jay@gnu.org>
 */


/* $XConsortium: lndir.c,v 1.13 94/04/17 20:10:42 rws Exp $ */
/* Create shadow link tree (after X11R4 script of the same name)
   Mark Reinhold (mbr@lcs.mit.edu)/3 January 1990 */

/*
Copyright (c) 1990,  X Consortium

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of the X Consortium shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from the X Consortium.

*/

/* From the original /bin/sh script:

  Used to create a copy of the a directory tree that has links for all
  non-directories (except those named RCS, SCCS or CVS.adm).  If you are
  building the distribution on more than one machine, you should use
  this technique.

  If your master sources are located in /usr/local/src/X and you would like
  your link tree to be in /usr/local/src/new-X, do the following:

        %  mkdir /usr/local/src/new-X
        %  cd /usr/local/src/new-X
        %  lndir ../X
*/
#include <config.h>

#ifndef HAVE_SYMLINK
#error I need to be patched to support either hard links or copying.
#endif

#ifndef HAVE_READLINK
#error I need to be patched to support either hard links or copying.
#endif


#include <locale.h>
#include <stdlib.h>
#include <assert.h>
#include <stdarg.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/types.h>

#include "dirent-safer.h"
#include "progname.h"
#include "gettext.h"

#ifndef MAXPATHLEN
#define MAXPATHLEN 2048
#endif

int silent;

char *rcurdir;
char *curdir;

static void vmsg (const char *, int, va_list);

void
quit (int code, int err_num, char * fmt, ...)
{
    va_list args;
    va_start(args, fmt);
    vmsg (fmt, err_num, args);
    va_end(args);
    exit (code);
}

static void
vmsg (const char *fmt, int errnum, va_list ap)
{
  assert (fmt && fmt[0]);
  if (curdir) {
    fprintf (stderr, "%s:\n", curdir);
    curdir = 0;
  }
  vfprintf (stderr, fmt, ap);
  if (errnum)
    {
      fprintf (stderr, ": %s", strerror (errnum));
    }
  putc ('\n', stderr);
}

void
mperror (int err_num, const char *fmt, ...)
{
  va_list args;
  va_start(args, fmt);
  vmsg (fmt, err_num, args);
  va_end(args);
}

int equivalent(char *lname, const char *rname)
{
    char *s;

    if (!strcmp(lname, rname))
        return 1;
    for (s = lname; *s && (s = strchr(s, '/')); s++) {
        while (s[1] == '/')
            strcpy(s+1, s+2);
    }
    return !strcmp(lname, rname);
}

static int is_dir(const struct stat *p)
{
  return S_ISDIR(p->st_mode);
}

static int ignored(const char *n)
{
  const char *histdirs[] = {"SCCS","RCS","CVS.adm","CVS", (const char*)0};
  const char **p;
  for (p=histdirs; *p; ++p)
    {
      if (0 == strcmp(*p, n))
        return 1;
    }
  return 0;
}

static int
is_dot_or_dotdot (const char *p)
{
  if (p[0] == '.') {
    if (p[1] == 0) {
      return 1;			/* . */
    } else if (p[1] == '.' && p[2] == 0) {
      return 1;			/* .. */
    }
  }
  return 0;
}


/* Recursively create symbolic links from the current directory to the "from"
   directory.  Assumes that files described by fs and ts are directories. */

int
dodir (const char *fn, /* name of "from" directory, either absolute or relative to cwd */
       const struct stat *fs, /* stats for the "from" directory */
       const struct stat *ts, /* stats for the cwd */
       int rel)
{
    DIR *df;
    struct dirent *dp;
    char buf[MAXPATHLEN + 1], *p;
    char symbuf[MAXPATHLEN + 1];
    struct stat sb, sc;
    int symlen;
    char *ocurdir;

    if ((fs->st_dev == ts->st_dev) && (fs->st_ino == ts->st_ino))
      {
        mperror (errno,
		 "%s: From and to directories are identical, "
		 "hence no work to do!", fn);
        return 0;               /* nothing to do! */
      }

    if (rel)
        strcpy (buf, "../");
    else
        buf[0] = '\0';
    strcat (buf, fn);

    if (!(df = opendir (buf))) {
        mperror (errno, "%s: Cannot opendir", buf);
        return 1;
    }

    p = buf + strlen (buf);
    *p++ = '/';

    while ( (struct dirent*)0 != (dp = readdir (df)) ) {
        const size_t namlen = strlen(dp->d_name);
	if (namlen) {
	  if (dp->d_name[namlen - 1] == '~')
            continue;
	}
	if (is_dot_or_dotdot (dp->d_name)) {
	  continue;		/* ignore. */
	}


        strcpy (p, dp->d_name);

        if (1) {
            if (stat (buf, &sb) < 0) {
	        mperror (errno, "failed to stat %s", buf);
                continue;
            }


            if (is_dir(&sb))
            {
                /* directory */
                if (is_dot_or_dotdot (dp->d_name) || ignored (dp->d_name))
                    continue;

                ocurdir = rcurdir;
                rcurdir = buf;
                curdir = silent ? buf : (char *)0;
                if (!silent)
                    printf ("making links in %s:\n", buf);
                if ((stat (dp->d_name, &sc) < 0) && (errno == ENOENT)) {
                    if (mkdir (dp->d_name, 0777) < 0 ||
                        stat (dp->d_name, &sc) < 0) {
		        mperror (errno, "failed to stat %s", dp->d_name);
                        curdir = rcurdir = ocurdir;
                        continue;
                    }
                }
                if (readlink (dp->d_name, symbuf, sizeof(symbuf) - 1) >= 0) {
		    mperror (0, "%s: is a link instead of a directory",
			     dp->d_name);
                    curdir = rcurdir = ocurdir;
                    continue;
                }
                if (chdir (dp->d_name) < 0) {
		    mperror (errno,
			     "failed to change directory into %s", dp->d_name);
                    curdir = rcurdir = ocurdir;
                    continue;
                }
                dodir (buf, &sb, &sc, (buf[0] != '/'));
                if (chdir ("..") < 0)
		  quit (1, errno, "..");
                curdir = rcurdir = ocurdir;
                continue;
            }
        }

        /* non-directory */
        symlen = readlink (dp->d_name, symbuf, sizeof(symbuf) - 1);
        if (symlen >= 0) {
            symbuf[symlen] = '\0';
            if (!equivalent (symbuf, buf))
                mperror (0, "%s: %s", dp->d_name, symbuf);
        } else if (symlink (buf, dp->d_name) < 0)
            mperror (errno,
		     "failed to create a symbolic link %s pointing to %s",
		     dp->d_name, buf);
    }

    closedir (df);
    return 0;
}

int
main (int ac, char *av[])
{
    char *fn, *tn;
    struct stat fs, ts;

    set_program_name (av[0]);
    if (NULL == setlocale(LC_ALL, ""))
      {
	/* If we can't set the locale as the user wishes,
	 * emit an error message and continue.   The error
	 * message will of course be in the "C" locale.
	 */
	perror("Error setting locale");
      }
    bindtextdomain (PACKAGE, LOCALEDIR);
    textdomain (PACKAGE);

    silent = 0;
    if (ac > 1)
      {
        if (!strcmp(av[1], "--silent")) /* GNU-style long options. */
          silent = 1;
        else if (!strcmp(av[1], "-silent")) /* X11R4 compatibility. */
          silent = 1;
      }

    if (ac < silent + 2 || ac > silent + 3)
        quit (1, 0, "usage: %s [-silent] fromdir [todir]", av[0]);

    fn = av[silent + 1];
    if (ac == silent + 3)
        tn = av[silent + 2];
    else
        tn = ".";

    /* to directory */
    if (stat (tn, &ts) < 0)
        quit (1, errno, "%s", tn);

    if (!is_dir(&ts))
        quit (2, 0, "%s: Not a directory", tn);

    if (chdir (tn) < 0)
        quit (1, errno, "%s", tn);

    /* from directory */
    if (stat (fn, &fs) < 0)
        quit (1, errno, "%s", fn);

    if (!is_dir(&fs))
        quit (2, 0, "%s: Not a directory", fn);

    return dodir (fn, &fs, &ts, 0);
}
